### ISSUE TITLE
NullPointerException when training CRF

### ISSUE THREAD

**Issue by opaetzel** [haifengl/smile#218]
I instantiated a CRF.Trainer like this:
`CRF.Trainer trainer = new CRF.Trainer(6, 2);`
and tried to train a CRF with these [features](http://oliverpaetzel.de/features.json.gz) and [labels](http://oliverpaetzel.de/labels.json.gz). It fails with the following stack trace:

> Exception in thread "main" java.lang.NullPointerException  
at smile.sequence.CRF$Trainer$BoostingTask.<init>(CRF.java:706)  
at smile.sequence.CRF$Trainer.train(CRF.java:570)

**Comment by haifengl** [comment]
Can you give more code snippets on how you call CRF?

**Comment by opaetzel** [comment]
I call it this way, the featureArr and labelArr are the linked json data as arrays:
`CRF classifier = trainer.train(featureArr, labelArr);`

**Comment by haifengl** [comment]
I am afraid that featureArr or labelArr may not be in the expected way. Can you show how you create these variables.

**Comment by opaetzel** [comment]
you could download the files linked above and use [gson](https://github.com/google/gson) to deserialize them: 
``` java
double[][][] featureArr = null;
try (InputStream is = Files.newInputStream(Paths.get("/path/to/features.json.gz")); GZIPInputStream gis =
                            new GZIPInputStream(is); Reader jsonReader = new InputStreamReader(gis)) {
    featureArr = new Gson().fromJson(jsonReader, double[][][].class);
 }
int[][] labelArr = null;
try (InputStream is = Files.newInputStream(Paths.get("/path/to/labels.json.gz")); GZIPInputStream gis =
                            new GZIPInputStream(is); Reader jsonReader = new InputStreamReader(gis)) {
    labelArr = new Gson().fromJson(jsonReader, int[][].class);
 }
```

**Comment by haifengl** [comment]
I cannot download your data right now. But can you check if your data meet our assumptions:

the features are in [0, 5] and class labels in 0 or 1 for your data.

Thanks!

**Comment by opaetzel** [comment]
the features are normalized in [-1,1]. That could be the problem, if it has to be in [0,5]. Would normalizing to [0,1] help?

**Comment by haifengl** [comment]
This CRF implementation supposes working on sparse binary data (mostly for NLP).  See

https://github.com/haifengl/smile/blob/master/core/src/test/java/smile/sequence/CRFTest.java

how it works. And

https://github.com/haifengl/smile/blob/master/shell/src/universal/data/sequence/sparse.protein.11.train

is a sample data.

**Comment by opaetzel** [comment]
Ah okay I am using it wrong then. I am feeding it this data (extract):
``` json
[
    [
        [
            0.9502296043498872,
            0.04901466986045237,
            0.0,
            0.034034610957892716,
            1.0,
            0.013876651982378854
        ]
    ],
    [
        [
            0.4854410007187256,
            0.4864551706736476,
            0.0,
            0.017287421438929636,
            0.0,
            0.007048458149779736
        ],
        [
            0.24547584143671364,
            0.24372600251989104,
            0.08628030791852799,
            0.043218553597324086,
            0.0,
            0.024669603524229075
        ],
        [
            0.39752057179598427,
            0.3935003660316032,
            0.03328320483504048,
            0.01998858103876239,
            0.0,
            0.02378854625550661
        ],
        [
            0.4889477984535358,
            0.48720279644824843,
            0.0,
            0.012425334159230675,
            0.0,
            0.021145374449339206
        ],
        [
            0.23971467372952548,
            0.23724657914001662,
            0.037891648581430695,
            0.033021676107955436,
            0.0,
            0.0026431718061674008
        ],
```
So it is not sparse at all and the features are actual doubles. But good to know how to use it correctly, for other projects. Do you know of a sequence tagging algorithm that can handle this data?

**Comment by haifengl** [comment]
Sorry, I don't know.
